woocommerce/woocommerce
# Product Collection: Trigger `wc-blocks_viewed_product` JS event

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds `wc-blocks_viewed_product` event being triggered from three places:
- Product Image
- Product Title
- Add to Cart when it's an anchor (e.g. "Select Options", "Read More", etc.)

I used different strategies to implement the action:
- Product Image: I added Interactivity API and `view.ts` file with the action to trigger event on click
- Product Title: we use a Post Title variation. Hence I used Product Collection (which introduces this variation with namespace) as a place to insert directives through `render_block_core/post-title` filter. It's a similar pattern as for pagination and injecting client side navigation. It uses Product Collection namespace and context as it's not going to be used anywhere else.
- Add to Cart: I simply extended existing solution. 

Closes https://github.com/woocommerce/woocommerce/issues/48861



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. Use some plugin to include custom JS script on the page e.g. [Easy Code Snippets](https://wordpress.org/plugins/easy-code-snippets/)
2. Attach this code:

```
<script>
document.body.addEventListener(
	'wc-blocks_viewed_product',
	e => console.log('Event triggered', e.detail)
);
</script>
```

If you're using Easy Code Snippets, don't forget to wrap it in `<script></script>` tag and set the position to "Footer":
<img width="1159" alt="image" src="https://github.com/user-attachments/assets/c6a726f0-b204-4770-aa70-c6499a92b0ec">

#### Testing
1. Make sure you have some variable, grouped and out of stock products
2. Create new post
3. Add Product Collection block
4. Choose "create your own" so it shows all of the products
4. Save and go to frontend
5. Open browser's console
6. Make sure you enable "Preserve log" option (example view from Chrome browser, but it may be in different place in your browser):

![Screen Shot 2024-08-28 at 10 15 31 AM](https://github.com/user-attachments/assets/63e9b7d5-701a-457b-90ad-a7bd4b729bec)

8. For variable, grouped, and out-of-stock products click their "Add to Cart button" which should have labels like
    - "Select options"
    - "View products"
    - "Read more"
9. **Expected:** Along with redirection there should be an event logged in your console. Event should look like this with the `productId` corresponding to the product you clicked:

<img width="335" alt="image" src="https://github.com/user-attachments/assets/c57f7fa5-aa05-4505-a80a-960f72b59537">



### Changelog entry








<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
